### ISSUE TITLE
Internal Layer Referencing 1 - Outputs referencing inputs

### ISSUE THREAD

**Issue by joewheaton** [Riverscapes/riverscapes-tools#596]
### Internal Layer Referencing 1 - Every intermediate and output within a project should refer to the input(s) or intermediate(s) within that project that it was derived from.

I am not convinced that anyone heard or saw this ... @MattReimer says in #482 that this will be covered with the new project references. My understanding of project-referencing is that it is referencing of layers or datasets between projects.  This is about leaving metadata breadcrumbs (something the Tool Developer - i.e. @philipbaileynar, @KellyMWhitehead , @jtgilbert  has to proactively do).  I am not so sure. I don't see how this happens without a META tag being added to the `project.rs.xml` for every intermediate and every output, that simply states which inputs or intermediates it was derived from and if any parameters (e.g. passed internally in the code or config file) are stored as metadata in that part of the project. 

### Example
@MattReimer will no doubt have a better way of saying this, but in my simpleton view here is an example of what an output would need to do in a VBET project. 

The main VBET Evidence Raster has a listing like 
```xml
<Outputs>
  <Raster guid="f5f88612-1870-4ce3-96f1-ca3f164fcbd6" id="VBET_EVIDENCE">
    <Name>VBET Evidence Raster</Name>
    <Path>outputs/VBET_Evidence.tif</Path>
  </Raster>
</Outputs>
```

I presume we would need to add something like he's already done for cross-project references back on the inputs after the `<Path></Path>`, but before the `</Raster>` to say that it was based on the Normalized HAND, Normalized SLOPE, and Normalized Topo intermediates (which in this file have `id`s of `id="NORMALIZED_SLOPE"`,  `id="NORMALIZED_HAND"` and `id="NORMALIZED_TWI"`. 

```xml
<MetaData>
  <Meta name="_project_derived_from">NORMALIZED_SLOPE</Meta>
  <Meta name="_project_derived_from">NORMALIZED_HAND</Meta>
  <Meta name="_project_derived_from">NORMALIZED_TWI</Meta>
</MetaData>
```
Now, you can all say you're off your rocker @joewheaton. But this is **how** we get to those fundamental questions posed in https://github.com/Riverscapes/rs-web-monorepo/discussions/292.  Specifically, **what was this layer calculated from**? My argument is that a consumer of the project in RAVE should not have to dig into the source code to know exactly what went into producing a single layer I am looking at. Now, If I go back to the Normalized, HAND raster, I would chase that back to a HAND input, and realize it came from TauDEM (i.e. a "Cross-Project Reference" instead of "Internal Project Reference") and I can keep tracing it back to its origin and raw data. This is the essence of transparency as well as the *R* in FAIR (https://github.com/Riverscapes/rs-web-monorepo/discussions/285). 

So the first part of this is having the standard for how to make these links. The next is the developers following it in writing production grade tools. The last is that RAVE recognizes and exposes it on a right click in the tree and lets user look at those. 
----------
I am assigning it to @joewheaton until I see the Alpha release of the Warehouse 3.X with data in it. Once I have confirmed it is there I will close it. Otherwise, this will require @MattReimer to propose a Project XML Data tag standard for how we do it, and then each of our production grade models

- RS Context will have none of these (i.e. because it does not calculate anything, it just packages existing datasets).
- [ ] TauDEM will need to be modified to produce Internal Layer Referencing
- [ ] VBET will need to be modified to produce Internal Layer Referencing 
- [ ] ChannelArea  will need to be modified to produce Internal Layer Referencing 
- [ ] BRAT  will need to be modified to produce Internal Layer Referencing 
- [ ] RCAT will need to be developed to produce Internal Layer Referencing

**Comment by MattReimer** [comment]
This conversation is important I think and part of a much larger one about how opinionated we want/need to be about our project XMLs.

The chief problem here is that we don't really control the shape of metadata in the system. The solution you're proposing above is only a suggestion that all tool developers will have to choose to follow and we have no real way of checking that they've done it the same way as everyone else. 

This is not a problem really. I can almost hear the responses now of "we don't care what people do... let them be free!!!" but if we're talking about having a feature in the warehouse that can extract this information then we very much need a way to control it. 

The second problem I see is that the spread of tool functionality is so wide that getting any 2 of them to agree on anything is hard. In fact, we can't even really agree on what constitutes an "input" vs an "intermediate" dataset. I see this as being at best a voluntary feature that tool developers *can* use if they want the warehouse to operate. 

TL;DR (which normally belongs at the top so sorry about that): This is part of a bigger conversation about referencing datasets vs projects and we need to have it before I can suggest a fix of any kind.

**Comment by joewheaton** [comment]
I am with you 100% on this being a _voluntary_ feature that developers can optionally implement. I am saying:

1. I want every tool that we (NAR/USU) develop to follow this optional feature standard.
2. I want you to define the standard for how we do it.
3. I fully accept that the functionality and features will only work if the developer bothered to do it.

We can and do lead by example, but this is not all or nothing. You heard it right "we don't care what people do... let them be free" is exactly what I would say rhetorically. But I suppose what I really mean is "we don't need to lock people in to following every standard we set, but we can provide examples of higher standards that they can choose to adopt or follow if it suits them, and if they do they get a functional gain." More specifically, I selfishly want the functional gain for our tools.  I want a developer outside our bubble to notice that feature and ask, "how do I get my model to do that?".

**Comment by philipbaileynar** [comment]
TLDR: Data flows are tool specific. Accomplish this with [workflow diagrams](https://docs.google.com/drawings/d/1EMXc4DTrzHDOiNvKj125WDqpB32UiMrNTaK6ThNT76E/edit?usp=sharing) and better documentation, not metadata tags.

This needs careful discussion before anyone tackles this. My interpretation of this request is that: "every intermediate and output layer should reference the layers that were used to generate it". 

I think trying to accomplish this in metadata is the wrong approach. The dimensionality of our tools are too varied. For example in BRAT, there is one output layer. Inputs? EVERYTHING! This isn't useful to anyone!

@joewheaton states "people shouldn't have to read the source code" and I agree. But they should have to understand the tool workflow using good documentation and [diagrams like we have for VBET](https://docs.google.com/drawings/d/1EMXc4DTrzHDOiNvKj125WDqpB32UiMrNTaK6ThNT76E/edit?usp=sharing). Deducing from metadata is too fiddly and doesn't work uniformly for all our tools (see BRAT example where what is really needed is documenting which inputs feed into which attribute columns in the one output).

**Comment by joewheaton** [comment]
@MattReimer is this something we can close? Or is it something that needs more tool-specific tasks created for to confirm latest flavors are referencing correctly?

**Comment by MattReimer** [comment]
Yeah. reading through this again it seems like there's nothing actionable beyond the diagraming you're already doing. Let's close it.
